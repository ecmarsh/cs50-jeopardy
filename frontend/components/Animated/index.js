import Ease from './_Ease';
import Fade from './_Fade';
import Grow from './_Grow';
import Loud from './_Loud';
import Shift from './_Shift';
import Slide from './_Slide';
import Tally from './_Tally';
import Tree from './_Tree';
import Unfocus from './_Unfocus';

export { Ease, Fade, Grow, Loud, Shift, Slide, Tally, Tree, Unfocus };
