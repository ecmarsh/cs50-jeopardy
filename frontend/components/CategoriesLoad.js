import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { categoriesPerPage } from '../config';
import CategoryQuestions from './CategoryQuestions';
import { Heading } from './Views';
import Pagination from './Pagination';
import Error from './ErrorMessage';
import Main from './styles/Main';
import { Divider } from './styles/Utilities';

const CATEGORIES_QUERY = gql`
  query CATEGORIES_QUERY($gameName: String, $first: Int, $skip: Int) {
    categories(
      where: { game: { name: $gameName } }
      first: $first
      skip: $skip
    ) {
      id
      name
    }
  }
`;

class LoadCategories extends Component {
  state = {};

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { game, round, headingProps } = this.props;
    return (
      <Query
        query={CATEGORIES_QUERY}
        variables={{
          gameName: game,
          first: categoriesPerPage,
          skip: categoriesPerPage * (round - 1),
        }}
      >
        {({ data, loading, error }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <Error error={error} />;
          if (!data.categories) {
            return <Heading game={game} {...headingProps} />;
          }
          return (
            <Main>
              <Heading game={game} {...headingProps} />
              <Divider style={{ margin: '0 auto' }} />
              <Pagination gameName={game} round={round} />
              <div className="canvas">
                {data.categories.map(category => (
                  <div
                    key={'category' + category.id}
                    className="column-wrapper"
                  >
                    <ul>
                      <li>{category.name}</li>
                      <CategoryQuestions
                        categoryId={category.id}
                        gameName={game}
                        round={round}
                      />
                    </ul>
                  </div>
                ))}
              </div>
            </Main>
          );
        }}
      </Query>
    );
  }
}

export default LoadCategories;
export { CATEGORIES_QUERY };
