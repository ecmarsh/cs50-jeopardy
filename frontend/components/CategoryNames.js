import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import * as R from 'ramda';
import formatRound from '../lib/formatRound';
import Error from './ErrorMessage';
import Form from './styles/Form';
import Button from './styles/Button';
import { TextMuted, Divider } from './styles/Utilities';

const UPDATE_CATEGORY_MUTATION = gql`
  mutation UPDATE_CATEGORY_MUTATION($categoryId: ID!, $categoryName: String) {
    updateCategory(id: $categoryId, name: $categoryName) {
      id
      name
    }
  }
`;

class CategoryNames extends Component {
  static propTypes = {
    categories: PropTypes.array.isRequired,
  };

  state = {
    categories: [],
    called: false,
  };

  componentDidMount() {
    const { categories } = this.props;
    this.setState({ categories });
  }

  handleChange = e => {
    e.preventDefault();
    const { id, value } = e.target;
    const categories = [...this.state.categories];
    const idx = R.findIndex(R.propEq('id', id))(categories);
    categories[idx].name = value;
    categories[idx].updated = true; // flag to include in mutation
    this.setState({ categories });
  };

  handleSubmit = (e, updateCategoryMutation) => {
    // Stop default form submission
    e.preventDefault();
    // Get list of changed categories to update
    const updatedCategories = R.filter(R.has('updated'), this.state.categories);
    // Update category mutation
    const doMutation = async category => {
      const { id, name } = category;
      const res = await updateCategoryMutation({
        variables: { categoryId: id, categoryName: name },
      });
    };
    // Invoke mutation on each updated category
    R.forEach(doMutation, updatedCategories);
    // Called to flash saved message
    this.setState({ called: true });

    setTimeout(this.flashUpdatedConfirmation, 2000);
  };

  flashUpdatedConfirmation = () => {
    this.setState({ called: false });
  };

  render() {
    const { categories, called } = this.state;
    return (
      <Mutation mutation={UPDATE_CATEGORY_MUTATION}>
        {(updateCategory, { loading, error }) => {
          return (
            <Form
              data-test="category-names-form"
              className="noBg"
              onSubmit={e => this.handleSubmit(e, updateCategory)}
            >
              <h2>Category Names</h2>
              <Error error={error} />
              <fieldset disabled={loading} aria-busy={loading}>
                {categories.map((category, i) => (
                  <div key={category.id}>
                    {i === 0 && <h4>{formatRound(1)}</h4>}
                    {i === 5 && <h4>{formatRound(2)}</h4>}
                    {i === 10 && <h4>{formatRound(3)}</h4>}
                    <label
                      htmlFor={`category-${i}`}
                      style={{
                        flexDirection: 'row',
                        width: '100%',
                      }}
                    >
                      <TextMuted
                        style={{ textAlign: 'center', flexBasis: '25%' }}
                      >
                        CATEGORY {i + 1}:
                      </TextMuted>
                      <input
                        type="text"
                        name={`category-${i}`}
                        id={category.id}
                        value={category.name}
                        onChange={e => this.handleChange(e)}
                        style={{
                          flexBasis: '60%',
                          width: 'auto',
                        }}
                      />
                    </label>
                  </div>
                ))}
                <div style={{ margin: '0 auto' }}>
                  <TextMuted
                    data-test="category-names-save-confirmation"
                    hidden={!called}
                  >
                    Category Names Saved!
                  </TextMuted>
                  <Button
                    primary
                    type="submit"
                    style={{ margin: '1rem auto 2rem' }}
                  >
                    Save Category Names
                  </Button>
                  <Divider />
                  {this.props.children}
                </div>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default CategoryNames;
export { UPDATE_CATEGORY_MUTATION };
