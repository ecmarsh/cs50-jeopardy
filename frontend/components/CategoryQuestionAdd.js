import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import DropTarget from './styles/DropTarget';
import Error from './ErrorMessage';
import RemoveFromCategory from './CategoryQuestionRemove';
import { GAME_QUESTIONS_QUERY } from './QuestionsList';
import { CATEGORY_QUESTIONS_QUERY } from './CategoryQuestions';
import { GAME_SUMMARY_QUERY } from './GameSummary';
import * as R from 'ramda';

const ADD_TO_CATEGORY_MUTATION = gql`
  mutation ADD_TO_CATEGORY_MUTATION($id: ID!, $questionId: ID!) {
    updateCategoryQuestion(id: $id, questionId: $questionId) {
      id
      question {
        categoryQuestions {
          id
        }
      }
    }
  }
`;

class AddToCategory extends Component {
  static propTypes = {
    loaded: PropTypes.bool,
    question: PropTypes.object,
    categoryQuestionId: PropTypes.string.isRequired,
    categoryId: PropTypes.string.isRequired,
    difficulty: PropTypes.number.isRequired,
    categoryName: PropTypes.string.isRequired,
    gameName: PropTypes.string.isRequired,
  };

  state = {
    questionId: String(),
    question: String(),
  };

  componentDidMount = () => {
    if (this.props.question) {
      const { id, question } = this.props.question;
      this.setState({
        questionId: id,
        question,
      });
    }
  };

  handleDragOver = e => {
    e.preventDefault();
    e.persist(); // don't pool the events
    e.target.style.transform = 'scale(1.2)';
    e.dataTransfer.dropEffect = 'move';
  };

  handleDragLeave = e => {
    e.target.style.transform = 'scale(1)';
  };

  handleDrop = async (e, categoryMutation) => {
    e.persist();
    e.target.style.transform = 'scale(1)';
    e.dataTransfer.dropEffect = 'move';

    const { categoryId, categoryQuestionId, difficulty, gameName } = this.props,
      questionId = e.dataTransfer.getData('id'),
      question = e.dataTransfer.getData('question'),
      answer = e.dataTransfer.getData('answer');

    this.setState({ questionId, question });

    const res = await categoryMutation({
      variables: { id: categoryQuestionId, questionId },
      update: (cache, { data: { updateCategoryQuestion } }) => {
        /* Update category question with added question */
        const categoryQuestionsQuery = {
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
        };
        // Read the data
        const { categoryQuestions } = cache.readQuery(categoryQuestionsQuery);
        // Add the question
        categoryQuestions[difficulty - 1].question = {
          answer,
          id: questionId,
          question,
          __typename: 'Question',
        };
        // Write back the cache
        cache.writeQuery({
          data: { categoryQuestions },
          ...categoryQuestionsQuery,
        });

        /* Update question to know it's in a category now */
        const gameQuestionsQuery = {
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
        };
        // Read data to questions variable
        const { questions } = cache.readQuery(gameQuestionsQuery);
        // Add new question into matching question's category list
        const idx = R.findIndex(R.propEq('id', questionId), questions);
        const questionToAdd = {
          id: updateCategoryQuestion.id,
          __typename: 'CategoryQuestion',
          category: {
            id: categoryId,
            __typename: 'Category',
          },
        };
        questions[idx].categoryQuestions.push(questionToAdd);
        // Write back cache with updates
        gameQuestionsQuery.data = { questions };
        //console.log(gameQuestionsQuery);
        cache.writeQuery(gameQuestionsQuery);
      },
    });
  };

  render() {
    const {
      loaded,
      question,
      categoryQuestionId,
      categoryId,
      difficulty,
      categoryName,
      gameName,
    } = this.props;

    return (
      <Mutation
        mutation={ADD_TO_CATEGORY_MUTATION}
        refetchQueries={[
          {
            query: GAME_SUMMARY_QUERY,
            variables: { gameName },
          },
        ]}
      >
        {(updateCategoryQuestion, { loading, error }) => {
          if (loading) return <DropTarget>Adding..</DropTarget>;
          if (error) return <Error error={error} />;
          return (
            <DropTarget
              id={`${categoryName}-${categoryId}`}
              onDragOver={!loaded ? this.handleDragOver : undefined}
              onDragLeave={!loaded ? this.handleDragLeave : undefined}
              onDrop={e => this.handleDrop(e, updateCategoryQuestion)}
              className={loaded ? '--isLoaded' : '--isNotLoaded'}
            >
              {loaded ? <div>{question.question}</div> : difficulty}
              {loaded && (
                <RemoveFromCategory
                  categoryQuestionId={categoryQuestionId}
                  categoryId={categoryId}
                  gameName={gameName}
                  difficulty={difficulty}
                  questionId={this.state.questionId}
                />
              )}
            </DropTarget>
          );
        }}
      </Mutation>
    );
  }
}

export default AddToCategory;
export { ADD_TO_CATEGORY_MUTATION };
