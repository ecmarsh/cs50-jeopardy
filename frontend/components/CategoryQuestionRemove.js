import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Button from './styles/Button';
import Error from './ErrorMessage';
import { CATEGORY_QUESTIONS_QUERY } from './CategoryQuestions';
import { GAME_QUESTIONS_QUERY } from './QuestionsList';
import { GAME_SUMMARY_QUERY } from './GameSummary';
import { TextLarger } from './styles/Utilities';
import * as R from 'ramda';

const REMOVE_FROM_CATEGORY_MUTATION = gql`
  mutation REMOVE_FROM_CATEGORY_MUTATION($id: ID!) {
    removeFromCategory(id: $id) {
      id
    }
  }
`;

class RemoveFromCategory extends Component {
  static propTypes = {
    children: PropTypes.string,
    categoryQuestionId: PropTypes.string.isRequired,
  };
  static defaultProps = {
    children: `âœ•`,
  };

  handleClick = async (e, removeFromCategoryMutation) => {
    e.preventDefault();
    const {
      questionId,
      categoryQuestionId,
      categoryId,
      difficulty,
      gameName,
    } = this.props;

    const res = await removeFromCategoryMutation({
      // Mutation variables
      variables: {
        id: categoryQuestionId,
      },
      // Update local data
      update: (cache, { data: { removeFromCategory } }) => {
        // Update the question so it knows it's not in a category
        const { questions } = cache.readQuery({
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
        });
        const questionIndex = questions.findIndex(
          question => question.id === questionId
        );

        const idx = R.findIndex(R.propEq('id', questionId));

        // Filter the category question out
        questions[questionIndex].categoryQuestions = questions[
          questionIndex
        ].categoryQuestions.filter(
          categoryQuestion => categoryQuestion.id !== removeFromCategory.id
        );

        // Write back the questions data
        cache.writeQuery({
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
          data: { questions },
        });
        // Read, update, writeback the loaded questions cache
        const { categoryQuestions } = cache.readQuery({
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
        });
        // Remove the question from the category
        categoryQuestions[difficulty - 1].question = null;
        // Write back the cache
        cache.writeQuery({
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
          data: { categoryQuestions },
        });
      },
      optimisticResponse: {
        __typename: 'Mutation',
        removeFromCategory: {
          id: categoryQuestionId,
          __typename: 'CategoryQuestion',
        },
      },
    });
  };

  render() {
    const { gameName } = this.props;
    return (
      <Mutation
        mutation={REMOVE_FROM_CATEGORY_MUTATION}
        refetchQueries={[
          {
            query: GAME_SUMMARY_QUERY,
            variables: { gameName },
          },
        ]}
      >
        {(removeFromCategory, { loading, error }) => {
          if (loading) return <p>Removing..</p>;
          if (error) return <Error error={error} />;
          return (
            <Button
              close
              onClick={e => this.handleClick(e, removeFromCategory)}
            >
              <TextLarger>{this.props.children}</TextLarger>
            </Button>
          );
        }}
      </Mutation>
    );
  }
}

export default RemoveFromCategory;
export { REMOVE_FROM_CATEGORY_MUTATION };
