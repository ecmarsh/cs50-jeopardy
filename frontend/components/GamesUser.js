import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import { format } from 'date-fns';
import { GAME_SUMMARY_QUERY } from './GameSummary';
import DeleteGame from './GameDelete';
import Error from './ErrorMessage';
import LeftColumn from './styles/List';
import StyledListItem from './styles/ListItem';
import { CloseSquareO } from './styles/icons';
import { TextOnOff } from './styles/Utilities';

const USER_GAMES_QUERY = gql`
  query USER_GAMES_QUERY {
    myGames {
      id
      name
      createdAt
      user {
        id
        name
      }
      questions {
        id
      }
    }
  }
`;

const GameStatus = ({ gameName }) => (
  <Query query={GAME_SUMMARY_QUERY} variables={{ gameName }}>
    {({ data: { categoryQuestionsConnection }, loading, error }) => {
      if (loading)
        return (
          <p>
            <small>Loading...</small>
          </p>
        );
      if (error)
        return (
          <p>
            <small>Questions Status</small>
          </p>
        );
      const loadedQuestions = categoryQuestionsConnection.aggregate.count;
      return (
        <p>
          <small>
            Question Status
            <TextOnOff textMuted={loadedQuestions < 51}>
              {loadedQuestions} / 51
            </TextOnOff>
          </small>
        </p>
      );
    }}
  </Query>
);
GameStatus.propTypes = {
  gameName: PropTypes.string.isRequired,
};

const UserGameListItem = ({ game: { id, name, createdAt } }) => (
  <StyledListItem key={id}>
    <Link
      href={{
        pathname: `/setup`,
        query: { game: name },
      }}
    >
      <a>
        <h4>{name}</h4>
        <GameStatus gameName={name} />
        <p>
          <small>{format(createdAt, 'MMMM DD, YYYY h:mm a')}</small>
        </p>
      </a>
    </Link>
    <DeleteGame gameName={name}>
      <CloseSquareO />
    </DeleteGame>
  </StyledListItem>
);
UserGameListItem.propTypes = {
  game: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    createdAt: PropTypes.string,
  }).isRequired,
};

const Heading = ({ isSolo }) => {
  function HeaderInner() {
    return (
      <Fragment>
        <div className="heading">
          <h3>My Games</h3>
          <Link href="/start">
            <a>+ Create New Game</a>
          </Link>
          <hr />
        </div>
      </Fragment>
    );
  }
  if (isSolo) {
    return (
      <LeftColumn>
        <HeaderInner />
      </LeftColumn>
    );
  }
  return <HeaderInner />;
};
Heading.propTypes = {
  isSolo: PropTypes.bool,
};

const UserGamesList = () => (
  <Query query={USER_GAMES_QUERY}>
    {({ data: { myGames }, loading, error }) => {
      if (loading) return 'Loading...';
      if (!myGames) return <Heading isSolo />;
      return (
        <LeftColumn>
          <Heading />
          <Error error={error} />
          <ul>
            {myGames.map(game => (
              <UserGameListItem key={game.id} game={game} />
            ))}
          </ul>
        </LeftColumn>
      );
    }}
  </Query>
);

export default UserGamesList;
export { USER_GAMES_QUERY, UserGameListItem, GameStatus };
