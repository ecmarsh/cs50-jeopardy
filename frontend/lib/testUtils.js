import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
