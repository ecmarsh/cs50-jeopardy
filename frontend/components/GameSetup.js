import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import CategoryNames from './CategoryNames';
import { Heading } from './Views';
import Main from './styles/Main';
import Button from './styles/Button';
import { Divider } from './styles/Utilities';

const SINGLE_GAME_QUERY = gql`
  query SINGLE_GAME_QUERY($name: String) {
    game(name: $name) {
      id
      name
      createdAt
      isPublic
      user {
        id
        name
        email
      }
      teams {
        id
        name
        score
      }
      categories {
        id
        name
        categoryQuestions {
          id
          answered
          difficulty
        }
      }
      config {
        id
        hasDoubleJeopardy
        hasRoundTimer
        roundTime
        finalTime
      }
    }
  }
`;

function GameSetup({ gameName }) {
  GameSetup.propTypes = { gameName: PropTypes.string.isRequired };
  return (
    <Query query={SINGLE_GAME_QUERY} variables={{ name: gameName }}>
      {({ data: { game }, loading, error }) => {
        if (loading) return <p>Loading...</p>;
        if (error)
          return (
            <Main>
              <h2>Name Your Categories</h2>
              <Divider />
              <h3>
                <Link href="/">
                  <a>{`Select`}</a>
                </Link>{' '}
                or{' '}
                <Link href="start">
                  <a>{`Create A Game`}</a>
                </Link>{' '}
                {`First ðŸ‘ˆ`}
              </h3>
            </Main>
          );
        return (
          <Main>
            <Heading
              game={gameName}
              title={gameName}
              prevLinkText="My Games"
              prevLinkPath="/"
              nextLinkPath="/load"
              nextLinkText="Load Questions"
            />
            <Divider />
            <CategoryNames categories={game.categories}>
              <Link
                prefetch
                href={{
                  pathname: `/load`,
                  query: { game: gameName },
                }}
              >
                <Button secondary type="button" style={{ margin: '1rem auto' }}>
                  Choose Questions &rarr;
                </Button>
              </Link>
            </CategoryNames>
          </Main>
        );
      }}
    </Query>
  );
}

export default GameSetup;
export { SINGLE_GAME_QUERY };
