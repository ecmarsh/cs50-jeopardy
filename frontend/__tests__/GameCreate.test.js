import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateGame, { CREATE_GAME_MUTATION } from '../components/GameCreate';
import { USER_GAMES_QUERY } from '../components/GamesUser';
import { SINGLE_GAME_QUERY } from '../components/GameSetup';
import Router from 'next/router';
import { fakeGame, fakeCategory } from '../lib/testUtils';

describe('<GameCreate />', () => {
  it('renders game create form and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateGame />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const form = wrapper.find('form[data-test="create-game-form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('updates state on input', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateGame />
      </MockedProvider>
    );
    wrapper.find('#gameName').simulate('change', {
      target: { value: 'test game', name: 'gameName' },
    });
    expect(wrapper.find('CreateGame').instance().state).toMatchObject({
      gameName: 'test game',
    });
  });

  it('does the game create mutation', async () => {
    const game = fakeGame();
    const cat = fakeCategory();
    const gameMock = {
      request: {
        query: CREATE_GAME_MUTATION,
        variables: { gameName: game.name },
      },
      result: {
        data: {
          createGame: {
            id: game.id,
            name: game.name,
            createdAt: game.createdAt,
            __typename: 'Game',
          },
          categoryOne: cat,
          categoryTwo: cat,
          categoryThree: cat,
          categoryFour: cat,
          categoryFive: cat,
          categorySix: cat,
          categorySeven: cat,
          categoryEight: cat,
          categoryNine: cat,
          categoryTen: cat,
          categoryFinal: cat,
        },
      },
    };
    const refetchMock = {
      request: {
        query: USER_GAMES_QUERY,
      },
      result: {
        data: {
          myGames: [game],
        },
      },
    };
    const gameQuery = {
      request: {
        query: SINGLE_GAME_QUERY,
      },
      result: {
        data: { game },
      },
    };
    const wrapper = mount(
      <MockedProvider mocks={[gameMock, refetchMock]} addTypename={false}>
        <CreateGame />
      </MockedProvider>
    );

    wrapper.find('#gameName').simulate('change', {
      target: { value: game.name, name: 'gameName' },
    });
    Router.router = { push: jest.fn() };
    wrapper.find('form[data-test="create-game-form"]').simulate('submit');
    await wait(50);
    wrapper.update();
    expect(Router.router.push).toHaveBeenCalled();
    const nextRoute = {
      pathname: '/setup',
      query: { game: game.name },
    };
    expect(Router.router.push).toHaveBeenCalledWith(nextRoute);
  });
});
