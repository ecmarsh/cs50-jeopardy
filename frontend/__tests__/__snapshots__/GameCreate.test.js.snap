// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameCreate /> renders game create form and matches snapshot 1`] = `
<form
  className="Form-qsq7s5-0 bLtDzK"
  data-test="create-game-form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <h2>
    Create New Game
  </h2>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="gameName"
    >
      <input
        autoComplete="off"
        id="gameName"
        name="gameName"
        onChange={[Function]}
        placeholder="Game Name"
        required={true}
        type="text"
        value=""
      />
      <Utilities__TextMuted>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__TextMuted-s11g5i-2",
                "isStatic": false,
                "lastClassName": "kHcKTC",
                "rules": Array [
                  "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                ],
              },
              "displayName": "Utilities__TextMuted",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__TextMuted-s11g5i-2",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="Utilities__TextMuted-s11g5i-2 bdrBqE"
          >
            Game Name
          </span>
        </StyledComponent>
      </Utilities__TextMuted>
    </label>
    <Button
      primary={true}
      type="submit"
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Button-sc-1f1w3gq-0",
              "isStatic": false,
              "lastClassName": "jSaPjy",
              "rules": Array [
                "font-family:",
                [Function],
                ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
                [Function],
                " auto 3px;}",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"❌\\";padding-left:3px;}",
                [Function],
                "}",
                [Function],
              ],
            },
            "displayName": "Button",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Button-sc-1f1w3gq-0",
            "target": "button",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        primary={true}
        type="submit"
      >
        <button
          className="Button-sc-1f1w3gq-0 jSaPjy"
          type="submit"
        >
          Create Game
        </button>
      </StyledComponent>
    </Button>
    <Utilities__TextMuted
      smaller={true}
      style={
        Object {
          "marginTop": "2.5rem",
        }
      }
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Utilities__TextMuted-s11g5i-2",
              "isStatic": false,
              "lastClassName": "kHcKTC",
              "rules": Array [
                "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
              ],
            },
            "displayName": "Utilities__TextMuted",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Utilities__TextMuted-s11g5i-2",
            "target": "span",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        smaller={true}
        style={
          Object {
            "marginTop": "2.5rem",
          }
        }
      >
        <span
          className="Utilities__TextMuted-s11g5i-2 kHcKTC"
          style={
            Object {
              "marginTop": "2.5rem",
            }
          }
        >
          Please note it may take a minute to setup your new game.
        </span>
      </StyledComponent>
    </Utilities__TextMuted>
  </fieldset>
</form>
`;
